cmd/cloud-node-manager/app/nodemanager.go

```diff
@@ -34,126 +30,95 @@ import (
+			healthzHandler.AddHealthChecker(check)
+			unsecuredMux := genericcontrollermanager.NewBaseHandler(&config.DebuggingConfiguration{}, healthzHandler)
+			handler := genericcontrollermanager.BuildHandlerChain(unsecuredMux, &c.Authorization, &c.Authentication)
+			// no need to handle stoppedCh returned by c.SecureServing.Serve because here is health checker.
+			if _, _, err := c.SecureServing.Serve(handler, 0, ctx.Done()); err != nil {
+				return err
+			}
 		}
-	}
-
-	run(context.TODO())
-	panic("unreachable")
-}
-
-// startControllers starts the cloud specific controller loops.
-func startControllers(ctx context.Context, c *cloudnodeconfig.Config, stopCh <-chan struct{}, healthzHandler *controllerhealthz.MutableHealthzHandler) error {
-	klog.V(1).Infof("Starting cloud-node-manager...")
-
-	// Start the CloudNodeController
-	nodeController := nodemanager.NewCloudNodeController(
-		c.NodeName,
-		c.SharedInformers.Core().V1().Nodes(),
-		// cloud node controller uses existing cluster role from node-controller
-		c.ClientBuilder.ClientOrDie("node-controller"),
-		nodeprovider.NewNodeProvider(ctx, c.UseInstanceMetadata, c.CloudConfigFilePath),
-		c.NodeStatusUpdateFrequency.Duration,
-		c.WaitForRoutes,
-		c.EnableDeprecatedBetaTopologyLabels)
 
-	go nodeController.Run(stopCh)
+		klog.V(1).Infof("Starting cloud-node-manager...")
+		var nodeProvider nodemanager.NodeProvider
 
-	check := controllerhealthz.NamedPingChecker(c.NodeName)
-	healthzHandler.AddHealthChecker(check)
-
-	klog.Infof("Started cloud-node-manager")
-
-	// If apiserver is not running we should wait for some time and fail only then. This is particularly
-	// important when we start node manager before apiserver starts.
-	if err := genericcontrollermanager.WaitForAPIServer(c.VersionedClient, 10*time.Second); err != nil {
-		klog.Fatalf("Failed to wait for apiserver being healthy: %v", err)
-	}
-
-	c.SharedInformers.Start(stopCh)
+		if c.UseInstanceMetadata {
+			nodeProvider = nodeprovider.NewIMDSNodeProvider(ctx)
+		} else {
+			nodeProvider = nodeprovider.NewARMNodeProvider(ctx, c.CloudConfigFilePath)
+		}
 
-	select {}
+		// Start the CloudNodeController
+		nodeController := nodemanager.NewCloudNodeController(
+			c.NodeName,
+			c.SharedInformers.Core().V1().Nodes(),
+			// cloud node controller uses existing cluster role from node-controller
+			c.ClientBuilder.ClientOrDie("node-controller"),
+			nodeProvider,
+			c.NodeStatusUpdateFrequency.Duration,
+			c.WaitForRoutes,
+			c.EnableDeprecatedBetaTopologyLabels)
+
+		// If apiserver is not running we should wait for some time and fail only then. This is particularly
+		// important when we start node manager before apiserver starts.
+		if err := genericcontrollermanager.WaitForAPIServer(c.VersionedClient, 10*time.Second); err != nil {
+			klog.Fatalf("Failed to wait for apiserver being healthy: %v", err)
+		}
+		c.SharedInformers.Start(ctx.Done())
+		go klog.Infof("Started cloud-node-manager")
```

Comment:
why use `go klog` here?

Reply:
nodecontroller.run will block the gorountine(143). starting a new routine to add log so that we don't need to wait(line 158)